og.api.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+/**
+ * @file
+ * Hooks provided by the Organic Groups module.
+ */
+
+declare(strict_types = 1);
+
+use Drupal\Core\Access\AccessResultInterface;
+use Drupal\Core\Cache\CacheableMetadata;
+use Drupal\Core\Entity\EntityPublishedInterface;
+use Drupal\og\OgAccess;
+
+/**
+ * @addtogroup hooks
+ * @{
+ */
+
+/**
+ * Allows modules to alter group level permissions.
+ *
+ * @param array $permissions
+ *   The list of group level permissions, passed by reference.
+ * @param \Drupal\Core\Cache\CacheableMetadata $cacheable_metadata
+ *   The cache metadata, passed by reference.
+ * @param array $context
+ *   An associative array containing contextual information, with keys:
+ *   - 'permission': The group level permission being checked, as a string.
+ *   - 'group': The group entity on which the permission applies.
+ *   - 'user': The user account for which access is being determined.
+ */
+function hook_og_user_access_alter(array &$permissions, CacheableMetadata &$cacheable_metadata, array $context): void {
+  // This example implements a use case where a custom module allows site
+  // builders to toggle a configuration setting that will prevent groups to be
+  // deleted if they are published.
+  // Retrieve the module configuration.
+  $config = \Drupal::config('mymodule.settings');
+
+  // Check if the site is configured to allow deletion of published groups.
+  $published_groups_can_be_deleted = \Drupal::config('mymodule')->get('delete_published_groups');
+
+  // If deletion is not allowed and the group is published, revoke the
+  // permission.
+  $group = $context['group'];
+  if ($group instanceof EntityPublishedInterface && !$group->isPublished() && !$published_groups_can_be_deleted) {
+    $key = array_search(OgAccess::DELETE_GROUP_PERMISSION, $permissions);
+    if ($key !== FALSE) {
+      unset($permissions[$key]);
+    }
+  }
+
+  // Since our access result depends on our custom module configuration, we need
+  // to add it to the cache metadata.
+  $cacheable_metadata->addCacheableDependency($config);
+}
+
+/**
+ * Allows to alter access to entity operations performed on group content.
+ *
+ * @param \Drupal\Core\Access\AccessResultInterface $access_result
+ *   The access result being altered.
+ * @param \Drupal\Core\Cache\CacheableMetadata $cacheable_metadata
+ *   The cache metadata, passed by reference.
+ * @param array $context
+ *   An associative array containing contextual information, with keys:
+ *   - 'operation': The entity operation being performed on the group content.
+ *   - 'group': The group entity to which the group content belongs.
+ *   - 'group_content': The group content entity upon which the operation is
+ *      performed.
+ *   - 'user': The user account for which access is being determined.
+ */
+function hook_og_user_access_entity_operation_alter(AccessResultInterface &$access_result, CacheableMetadata &$cacheable_metadata, array $context): void {
```

Comment:
Objects are passed by reference by default, not sure if it makes sense to explicitly do so here.

Reply:
I had the understanding that in PHP objects are not passed by reference, but rather you get a copy of the object instance, so you are working with a new object reference that points to the same memory location. If you then "shadow" the variable by instantiating it again then you lose the link with the original object. If you use a reference then your local variable will modify the original object instance.

So I think for the cache metadata we can remove the reference and force the user to work with the original object rather than replace it, but I think we need to keep the reference for the access result, since we need to be able to replace it with a new object.

I can be wrong though, I am nowadays more working with low level languages than with PHP and I'm possibly mixing things up. I will try this by removing the reference from the test and see if it fails.

Reply:
`So I think for the cache metadata we can remove the reference and force the user to work with the original object rather than replace it, but I think we need to keep the reference for the access result, since we need to be able to replace it with a new object.`

Indeed. I thought you only wanted to modify the object, if you want to replace the original one you do need the original reference to it.

> One of the key-points of PHP 5 OOP that is often mentioned is that "objects are passed by references by default". This is not completely true. This section rectifies that general thought using some examples.
> 
> A PHP reference is an alias, which allows two different variables to write to the same value. **As of PHP 5, an object variable doesn't contain the object itself as value anymore. It only contains an object identifier which allows object accessors to find the actual object. When an object is sent by argument, returned or assigned to another variable, the different variables are not aliases: they hold a copy of the identifier, which points to the same object.**

See https://www.php.net/manual/en/language.oop5.references.php

Reply:
I like this idea a lot, but then I think it is better to go with an event dispatcher right away. If we go with a normal hook then we have to provide a default implementation (like `og_og_user_access_entity_operation()`) otherwise we would have no way for modules to override a denied access originating from our code. With an event it is much easier to make OG's code run first and all other modules run after it.

Reply:
I gave this some thought and I will implement the following:

*AccessEventInterface*

An interface that can be used for all OG access events, so that it can be reused easily in the future for #686. This interface needs to supply common methods:

* Extend `RefineableCacheableDependencyInterface` to be able to supply cache metadata.
* Provide methods to deny and grant access. We will follow core's example that a single deny will overrule any grants.
* Provide methods to retrieve the group and user for which to check access. We can assume that in scope of OG there will always be a group in play when determining access for something.

*GroupContentAccessEventInterface*

This will extend `AccessEventInterface` and will provide additional methods to retrieve the operation and the group content entity.

*AccessEventBase*

Base class with shared methods for this PR + #686 

*GroupContentAccessEvent*

Implements the remaining methods of `GroupContentAccessEventInterface`

*OgEventSubscriber::checkGroupContentEntityOperationAccess()*

Will contain the logic that is now inside `OgAccess`

Reply:
Sounds like a lot of ~~work~~ fun :-)